### ISSUE TITLE
Add support to GORM

### ISSUE THREAD

**Issue by anduingaiden** [newrelic/go-agent#242]
[NOTE]: # ( ^^ Provide a general summary of the request in the title above. ^^ )

## Summary
Since many companies are using Golang each passing day, and most of them like to use some of the coolest libraries that the Go community develops, I believe that it would make sense for New Relic to support one of the most used DB ORM library, [GORM](https://gorm.io/).

## Additional context
This is could be a very useful thing for companies like mine, which uses New Relic to monitor its APIs made in GO.

**Comment by jodeev** [comment]
We are tracking Gorm support in the OpenTelemetry community. See https://github.com/open-telemetry/opentelemetry-go-contrib/issues/38

**Comment by RisingGeek** [comment]
How can I manually monitor database calls till the support for GORM is not added? I tried adding [Datastore segments](https://github.com/newrelic/go-agent/blob/master/GUIDE.md#datastore-segments) manually but with no success.

**Comment by songtianyi** [comment]
> How can I manually monitor database calls till the support for GORM is not added? I tried adding [Datastore segments](https://github.com/newrelic/go-agent/blob/master/GUIDE.md#datastore-segments) manually but with no success.

Any progress on this?

**Comment by Tarliton** [comment]
I was able to add GORM integration by creating a New Relic instrumented database connection (mysql in my case) .  
Then, I proceeded to create a GORM db instance from that existing database connection.

```go
db, err := sql.Open("nrmysql", "root@/information_schema")
if nil != err {
	panic(err)
}

gormdb, err := gorm.Open(mysql.New(mysql.Config{Conn: db}), &gorm.Config{})
if err != nil {
	panic("failed to connect database")
}
```

The tricky part is that we need create a new GORM DB instance for each new context. That's just how GORM works.
```go
gormdb = gormdb.WithContext(ctx)
```

![Untitled](https://user-images.githubusercontent.com/7471617/124787241-babea000-df1e-11eb-86c2-915920236683.png)
Full example [here](https://gist.github.com/Tarliton/ea9d42b14b0891ed28134a7b99d0e398).

**Comment by stephanebruckert** [comment]
@Tarliton your example works well, thanks! I think that's a better solution than manually adding callbacks to Gorm.

However it doesn't when adapted to postgres. This is likely an issue with the nrpq driver, see https://discuss.newrelic.com/t/go-datastore-segments-nrpq-not-working-when-transactions-in-general-are-working-fine/152573/2

**Comment by stephanebruckert** [comment]
I have created a new issue here https://github.com/newrelic/go-agent/issues/340

In the meantime, here is how you can integrate Gorm + Postgres with NR:

 - using gorm v1 + new-relic v1: https://github.com/smacker/newrelic-context
 - using gorm v1 + new-relic v3: https://github.com/filipemendespi/newrelic-context
 - using gorm v2 + new-relic v3: https://github.com/getndazn/newrelic-context (recommended)

It uses before/after callbacks on top of Gorm to send the right metrics to the NR database dashboards

![hi](https://user-images.githubusercontent.com/1932338/125294863-90088900-e31c-11eb-8325-2da49206b148.jpg)

**Comment by Ak-x** [comment]
@stephanebruckert
@songtianyi
@Tarliton
@RichVanderwal
@RisingGeek
@anduingaiden

Hello all, does the solution proposed by @Tarliton resolve this issue for everyone?

**Comment by stephanebruckert** [comment]
@Ak-x it did for me, it worked using Gorm through MySQL and Postgres

**Comment by VinitSarvade** [comment]
@Tarliton 
> The tricky part is that we need create a new GORM DB instance for each new context. That's just how GORM works.

When you mean new instance for each new context, do you mean for every API request we need to have a different instance?

**Comment by Tarliton** [comment]
> @Tarliton
> 
> > The tricky part is that we need create a new GORM DB instance for each new context. That's just how GORM works.
> 
> When you mean new instance for each new context, do you mean for every API request we need to have a different instance?

If by API request do you mean handling an incoming request, then yes.

**Comment by VinitSarvade** [comment]
@Tarliton Thank you so much. I will give it a shot.

**Comment by stephanebruckert** [comment]
Just to be clear, you don't want to create a new db/gorm connection on every request, but instead use the existing connection with a new transaction context.

So this will be done once:

```golang
db, err := sql.Open("nrmysql", "root@/information_schema")
gormdb, err := gorm.Open(mysql.New(mysql.Config{Conn: db}), &gorm.Config{})
```

But this should be done for every request:

```golang
txn := app.StartTransaction("mysqlQuery")
ctx := newrelic.NewContext(context.Background(), txn)
gormdb = gormdb.WithContext(ctx) 
```

(Referring to the gist example above)

**Comment by dogac00** [comment]
@stephanebruckert 

Do we have to name the current DB transaction like this?
``` go
txn := app.StartTransaction("mysqlQuery")
`
``

Isn't the purpose of the NewRelic database integrations not to give transaction names and have them inferred from the query?

**Comment by stephanebruckert** [comment]
@dogac00 ^ this is a minimal example

In my code I have a middleware where I name each transaction using the `requestUri` (endpoint)

```
func (a *App) newrelicMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		a.txn = a.newrelic.StartTransaction(cleanRequestUri(r))
		defer a.txn.End()

		r = r.WithContext(newrelic.NewContext(r.Context(), a.txn))

		a.txn.SetWebRequestHTTP(r)
		a.txn.SetWebResponse(w)

		next.ServeHTTP(w, r)
	})
}
```

which I use like this on the router (mux):

    r := mux.NewRouter().PathPrefix("/" + apiVersion).Subrouter().StrictSlash(true)
    r.Use(a.newrelicMiddleware)

**Comment by dogac00** [comment]
@stephanebruckert 

Hmm, I see.

So I think I can call `r.db.WithContext(ctx.Request().Context())` to get the underlying `context` from my `echo` request context which is `ctx` here, and then with that context I can use `db.WithContext` as we see in the code.

I have a similar middleware to you but with `echo` web framework;

https://github.com/newrelic/go-agent/blob/master/_integrations/nrecho/nrecho.go, here is my middleware.

I think since each request has different `context`s, `r.db.WithContext(ctx.Request().Context())` approach should work.

So I don't need to do `app.StartTransaction` and `newrelic.NewContext(context.Background(), txn)`, is that right?

**Comment by stephanebruckert** [comment]
> So I don't need to do `app.StartTransaction` and `newrelic.NewContext(context.Background(), txn)`, is that right?

That's right, it's already done in the middleware you are using.

Then yes as you said use `r.db.WithContext(ctx.Request().Context())` in each endpoint func, for example I do:

```golang
func (r *repo) ListAll(ctx context.Context) ([]*model.Foo, error) {
	var o []*model.Foo

	r.dbReader = r.dbReader.WithContext(ctx)

	err := r.dbReader.Find(&o).Error
	return o, err
}
```

**Comment by RichVanderwal** [comment]
We've just created [this issue](https://github.com/newrelic/go-agent/issues/438) to create documentation to help other customers with this as well.

**Comment by Ak-x** [comment]
This will be resolved via #438
